compiler/rustc_typeck/src/check/check.rs

```diff
@@ -426,8 +431,119 @@ fn check_struct(tcx: TyCtxt<'_>, def_id: LocalDefId, span: Span) {
     check_packed(tcx, span, def);
 }
 
+fn check_fields_for_opaque_types(
+    tcx: TyCtxt<'tcx>,
+    adt_def: &ty::AdtDef,
+    def_id: LocalDefId,
+    span: Span,
+) {
+    fn find_span_of_field_def_and_ty_alias(
+        tcx: TyCtxt<'tcx>,
+        field_def: &ty::FieldDef,
+    ) -> (Option<Span>, Option<Span>) {
+        let field_def_def_id = field_def.did;
+        if let Some(field_def_local_id) = field_def_def_id.as_local() {
+            let field_def_hir_id = tcx.hir().local_def_id_to_hir_id(field_def_local_id);
+            if let hir::Node::Field(hir::FieldDef {
+                span: field_def_span, ty: field_def_ty, ..
+            }) = tcx.hir().get(field_def_hir_id)
+            {
+                if let hir::TyKind::Path(hir::QPath::Resolved(_, path)) = field_def_ty.kind {
+                    if let Res::Def(DefKind::TyAlias, ty_alias_def_id) = path.res {
+                        if let Some(ty_alias_local_id) = ty_alias_def_id.as_local() {
+                            let ty_alias_hir_id =
+                                tcx.hir().local_def_id_to_hir_id(ty_alias_local_id);
+                            let node = tcx.hir().get(ty_alias_hir_id);
+                            match node {
+                                hir::Node::Item(hir::Item {
+                                    kind, span: ty_alias_span, ..
+                                }) => match kind {
+                                    hir::ItemKind::TyAlias(_, _) => {
+                                        return (Some(*field_def_span), Some(*ty_alias_span));
+                                    }
+                                    _ => bug!("expected an item of kind TyAlias"),
+                                },
+                                _ => return (Some(*field_def_span), None),
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        (None, None)
+    }
+
+    debug!("check_fields_of_opaque_types(adt_def: {:?}, span: {:?})", adt_def, span);
+
+    let item_type = tcx.type_of(def_id);
+    let substs = match item_type.kind() {
+        ty::Adt(_, substs) => substs,
+        _ => bug!("check_fields_for_opaque_types should only be called on Adts"),
+    };
+    adt_def.all_fields().for_each(|field_def| {
+        debug!("field_def: {:?}", field_def);
+
+        let field_ty = field_def.ty(tcx, substs);
+        match field_ty.kind() {
```

Comment:
```suggestion
        if field_ty.has_opaque_types() {
```
Would this be better? Might catch cases like `x: Vec<SomeOpaqueTypeAlias>` where the current wouldn't?